raft/src/main/java/org/apache/kafka/raft/internals/KafkaRaftMetrics.java

```diff
@@ -0,0 +1,215 @@
+        String metricGroupName = metricGrpPrefix + "-metrics";
+
+        this.pollStartMs = 0L;
+        this.pollEndMs = 0L;
+        this.electionStartMs = -1L;
+        this.numVoterConnections = 0;
+        this.logEndOffset = new OffsetAndEpoch(0L, 0);
+
+        this.currentStateMetricName = metrics.metricName("current-state", metricGroupName, "The current state of this member; possible values are leader, candidate, voter, observer");
+        Gauge<String> stateProvider = (mConfig, currentTimeMs) -> {
+            if (state.isLeader()) {
+                return "leader";
+            } else if (state.isCandidate()) {
+                return "candidate";
+            } else {
+                if (state.isVoter())
+                    return "voter";
+                else
+                    return "observer";
+            }
+        };
+        metrics.addMetric(this.currentStateMetricName, null, stateProvider);
+
+        this.currentLeaderIdMetricName = metrics.metricName("current-leader", metricGroupName, "The current quorum leader's id; -1 indicates unknown");
+        metrics.addMetric(this.currentLeaderIdMetricName, (mConfig, currentTimeMs) -> state.leaderId().orElse(-1));
+
+        this.currentVotedIdMetricName = metrics.metricName("current-vote", metricGroupName, "The current voted leader's id; -1 indicates not voted for anyone");
+        metrics.addMetric(this.currentVotedIdMetricName, (mConfig, currentTimeMs) -> {
+            if (state.isLeader() || state.isCandidate()) {
+                return state.localId;
+            } else {
+                FollowerState followerState = state.followerStateOrThrow();
+                if (followerState.hasVoted())
+                    return followerState.votedId();
+                else
+                    return -1;
+            }
+        });
+
+        this.currentEpochMetricName = metrics.metricName("current-epoch", metricGroupName, "The current quorum epoch.");
+        metrics.addMetric(this.currentEpochMetricName, (mConfig, currentTimeMs) -> state.epoch());
+
+        this.highWatermarkMetricName = metrics.metricName("high-watermark", metricGroupName, "The high watermark maintained on this member; -1 if it is unknown");
+        metrics.addMetric(this.highWatermarkMetricName, (mConfig, currentTimeMs) -> state.highWatermark().orElse(-1L));
+
+        this.logEndOffsetMetricName = metrics.metricName("log-end-offset", metricGroupName, "The current raft log end offset.");
+        metrics.addMetric(this.logEndOffsetMetricName, (mConfig, currentTimeMs) -> logEndOffset.offset);
+
+        this.logEndEpochMetricName = metrics.metricName("log-end-epoch", metricGroupName, "The current raft log end epoch.");
+        metrics.addMetric(this.logEndEpochMetricName, (mConfig, currentTimeMs) -> logEndOffset.epoch);
+
+        this.bootTimestampMetricName = metrics.metricName("boot-timestamp", metricGroupName, "The bootstrapped timestamp of this member.");
```

Comment:
What does this metric setup for? Couldn't we know the rough boot time from a dashboard, and this is only set once for the whole lifetime?

Reply:
Yes this is a static metric value, my intention is for trouble-shooting of phantom brokers. If we found it not useful I'm happy to remove it before the release.